Simplify build on macOS

0xced opened the pull request:
* Check for OpenSSL (and pkg-config) in the check_prereqs() function
* Simplify OpenSSL installation instructions
* Automatically pick the OpenSSL version installed by Homebrew by exporting the proper PKG_CONFIG_PATH environment variable
* Improve the error message if cmake can't find OpenSSL

The previous instructions were asking the user to add symbolic links inside  /usr/local/lib/ and /usr/local/lib/pkgconfig/ for OpenSSL related files. There is no need for a complicated setup with symbolic links when the PKG_CONFIG_PATH environment variable points to the right pkgconfig, which is now done in build-commons.sh.

0xced committed:
Simplify build on macOS

* Check for OpenSSL (and pkg-config) in the check_prereqs() function
* Simplify OpenSSL installation instructions
* Automatically pick the OpenSSL version installed by Homebrew by exporting the proper PKG_CONFIG_PATH environment variable
* Improve the error message if cmake can't find OpenSSL

The previous instructions were asking the user to add symbolic links inside  /usr/local/lib/ and /usr/local/lib/pkgconfig/ for OpenSSL related files. There is no need for a complicated setup with symbolic links when the PKG_CONFIG_PATH environment variable points to the right pkgconfig, which is now done in build-commons.sh.

0xced committed:
Order Homebrew packages alphabetically

0xced committed:
Remove `brew link --force icu4c` everywhere

It's not needed since everything is properly handled in the System.Globalization.Native CMakeLists.txt file: https://github.com/dotnet/runtime/blob/e301ec16723437561c84f7b9a8c773def67b81fd/src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt#L17

Anyway, recent versions of Homebrew refuse to link icu4c:

```sh
$ brew link --force icu4c
Warning: Refusing to link macOS provided/shadowed software: icu4c
If you need to have icu4c first in your PATH run:
  echo 'export PATH="/usr/local/opt/icu4c/bin:$PATH"' >> ~/.zshrc
  echo 'export PATH="/usr/local/opt/icu4c/sbin:$PATH"' >> ~/.zshrc

For compilers to find icu4c you may need to set:
  export LDFLAGS="-L/usr/local/opt/icu4c/lib"
  export CPPFLAGS="-I/usr/local/opt/icu4c/include"

For pkg-config to find icu4c you may need to set:
  export PKG_CONFIG_PATH="/usr/local/opt/icu4c/lib/pkgconfig"
```

jkotas commented on eng/native/build-commons.sh:
```diff
@@ -59,6 +59,19 @@ check_prereqs()
+
+        if ! pkg-config openssl ; then
+            # We export the proper PKG_CONFIG_PATH where openssl was installed by Homebrew
+            # Iit's important to _export_ it since build-commons.sh is sourced by other scripts such as build-native.sh
```
```suggestion
            # It's important to _export_ it since build-commons.sh is sourced by other scripts such as build-native.sh
```

0xced committed:
Fix typo

Co-Authored-By: Jan Kotas <jkotas@microsoft.com>

akoeplinger reviewed (approved):
This was bugging me for a long time, thanks for fixing it!

0xced committed:
Fix iOS and tvOS builds on Azure Pipelines

Use install-native-dependencies.sh instead of duplicating the list of native dependencies in the build job yaml files, just like it's done in other build jobs (eng/pipelines/common/templates/runtimes/build-test-job.yml, eng/pipelines/coreclr/templates/build-job.yml and eng/pipelines/mono/templates/build-job.yml)

akoeplinger commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -25,10 +25,10 @@ jobs:
       - script: |
           $(setScriptToEchoAndFailOnNonZero) 
-          brew install autoconf automake icu4c libtool openssl pkg-config python3
+          sh $(Build.SourcesDirectory)/eng/install-native-dependencies.sh $(osGroup)
```
looks like this needs to be `parameters.osGroup` and I think we can remove the `setScriptToEchoAndFailOnNonZero` to make it consistent with the other places.
```suggestion
          sh $(Build.SourcesDirectory)/eng/install-native-dependencies.sh ${{ parameters.osGroup }}
```

0xced replied:
I force-pushed 1956d5e1177e149b9d904f0c2175ae92fead7c42. But why is it different from the coreclr job? 🤯
https://github.com/dotnet/runtime/blob/89b2f425fadecd00fde47ad6a397e9065711ed61/eng/pipelines/coreclr/templates/build-job.yml#L107

akoeplinger replied:
I think that's because `osGroup` is part of a `variables` YML section there which is set to `${{ parameters.osGroup }}`.

akoeplinger commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -70,9 +70,8 @@ jobs:
-          - script: |
-              brew install icu4c openssl pkg-config
+        - ${{ if in(parameters.osGroup, 'OSX', 'iOS', 'tvOS') }}:
+          - script: sh $(Build.SourcesDirectory)/eng/install-native-dependencies.sh $(osGroup)
```
same here

```suggestion
          - script: sh $(Build.SourcesDirectory)/eng/install-native-dependencies.sh ${{ parameters.osGroup }}
```

0xced replied:
I force-pushed 1956d5e1177e149b9d904f0c2175ae92fead7c42.

akoeplinger commented:
Looks like this gets further now but `runtime-live-build` and `runtime (Installer Build and Test coreclr OSX_x64 Debug)` seem to not run the install-native-dependencies.sh script so they fail right now.

0xced committed:
Fix installer Azure pipeline

0xced commented:
I tried to fix the `runtime` pipeline wit 8f3eeb50692a092af7e10071084ac04ab0d95cb5 but I'm not sure if it's the best way.

The `runtime-live-build` pipeline fails with this error and I have no idea why and no idea how to diagnose this. 😕
> /Users/runner/runners/2.166.4/work/1/s/eng/install-native-dependencies.sh: No such file or directory

0xced committed:
Run install-native-dependencies.sh _after_ checkout

safern commented on eng/pipelines/installer/jobs/base-job.yml:
```diff
@@ -431,6 +431,10 @@ jobs:
         cleanUnpackFolder: false
 
+  - ${{ if in(parameters.osGroup, 'OSX', 'iOS', 'tvOS') }}:
+    - script: sh $(Build.SourcesDirectory)/eng/install-native-dependencies.sh ${{ parameters.osGroup }}
```
you don't need the `sh` do you?

wfurt replied:
*sh would be useful if you need particular shell flavor of if nstall-native-dependencies.sh  does not have executable bit.

0xced replied:
Addressed in 277c5b5ba66c44a0862a9a536e6aec32df747de9.

safern commented:
FYI: @bartonjs @wfurt

jashook commented on docs/workflow/requirements/macos-requirements.md:
```diff
@@ -23,47 +23,14 @@ Install the following packages:
 - automake
+- icu4c
 - libtool
+- openssl 1.1 or newer
```
Does newer work? I may have incorrectly thought we were stuck on 1.1. /cc @wfurt @janvorli

bartonjs replied:
We used to be stuck on 1.0.2; now we work anything 1.0.1-1.1.1 (though only 1.1.1 is still in support)

safern replied:
Yes it works. We no longer do dynamic linking. We just use it to build and we link at runtime against whatever is installed on the running machine. cc: @bartonjs

0xced replied:
I removed the *or newer* in f0f57416acc1cf78d2e96f2ed50492d1cb719e06. Let's be conservative about version 1.1.

safern commented on eng/native/build-commons.sh:
```diff
@@ -59,6 +59,19 @@ check_prereqs()
+            # It's important to _export_ it since build-commons.sh is sourced by other scripts such as build-native.sh
+            export PKG_CONFIG_PATH=/usr/local/opt/openssl/lib/pkgconfig
+            # We try again with the PKG_CONFIG_PATH in place, if pkg-config still can't find OpenSSL, exit with an error, cmake won't find OpenSSL either
+            pkg-config openssl || { echo >&2 "Please install openssl before running this script, see https://github.com/dotnet/runtime/blob/master/docs/workflow/requirements/macos-requirements.md"; exit 1; }
```
Should we check for other prereqs like `ICU`?

akoeplinger replied:
ICU won't be a system-wide prereq on some platforms like iOS though.

safern replied:
That's true... these messages are useful to the user though, maybe detect if targetOS is not iOS/Android/tvOS?

akoeplinger replied:
I don't think it's worth it, for ICU and other dependencies the usual cmake logic and error handling is straightforward.

safern replied:
Makes sense.

jashook reviewed (approved):
Minus the open question about openssl 1.1 lgtm

wfurt commented on docs/workflow/requirements/macos-requirements.md:
```diff
@@ -23,47 +23,14 @@ Install the following packages:
-ln -s /usr/local/opt/openssl\@1.1/lib/pkgconfig/libcrypto.pc /usr/local/lib/pkgconfig/
-ln -s /usr/local/opt/openssl\@1.1/lib/pkgconfig/libssl.pc /usr/local/lib/pkgconfig/
-ln -s /usr/local/opt/openssl\@1.1/lib/pkgconfig/openssl.pc /usr/local/lib/pkgconfig/
+brew install cmake autoconf automake icu4c libtool openssl pkg-config python3
```
Do you know if one can run tests locally without the links? (we would not see it the in CI as tests do run on different machine)

bartonjs replied:
Assuming Homebrew is still not linking the libraries into /usr/local/lib/, we need to keep the runtime dependencies section.

But as long as src\libraries\Native\build-native.sh is sourcing the script that adds the default homebrew path to the pkg-config probe path, we can remove the pkgconfig-specific steps from the docs.

wfurt replied:
If this is meant for developers, I'm wondering if we can add rpath so runtime would find them as well automatically. Or add note that somebody can set LD_LIBRARY_PATH as alternative to linking. 
In either case I would be nice if one does not have to do any linking and direct file manipulation.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native/CMakeLists.txt:
```diff
@@ -14,7 +14,7 @@ endif(CMAKE_STATIC_LIB_LINK)
 find_package(OpenSSL)
 if(NOT OPENSSL_FOUND)
-    message(FATAL_ERROR "!!! Cannot find libssl and System.Security.Cryptography.Native cannot build without it. Try installing libssl-dev (or the appropriate package for your platform) !!!. See the requirements document for your specific operating system: https://github.com/dotnet/runtime/tree/master/docs/workflow/requirements.")
+    message(FATAL_ERROR "!!! Cannot find libssl and System.Security.Cryptography.Native cannot build without it. Try installing libssl-dev (on Linux) or openssl (on macOS) !!!. See the requirements document for your specific operating system: https://github.com/dotnet/runtime/tree/master/docs/workflow/requirements.")
```
Not all distros use the same packaging; so the Linux name should still hedge.  And the macOS one might also be different from uh... MacPorts?  Whatever the Homebrew competitor is.

```suggestion
    message(FATAL_ERROR "!!! Cannot find libssl and System.Security.Cryptography.Native cannot build without it. Try installing libssl-dev (on Linux, but this may vary by distro) or openssl (on macOS) !!!. See the requirements document for your specific operating system: https://github.com/dotnet/runtime/tree/master/docs/workflow/requirements.")
```

wfurt commented:
Since you touching this, can you accommodate https://github.com/dotnet/runtime/issues/34150 ? Perhaps have `brew` as the primary option.

bartonjs commented on eng/native/build-commons.sh:
```diff
@@ -59,6 +59,19 @@ check_prereqs()
+        if ! pkg-config openssl ; then
+            # We export the proper PKG_CONFIG_PATH where openssl was installed by Homebrew
+            # It's important to _export_ it since build-commons.sh is sourced by other scripts such as build-native.sh
+            export PKG_CONFIG_PATH=/usr/local/opt/openssl/lib/pkgconfig
```
Presumably this should set the value additively, not as a replacement...

0xced replied:
I think replacement is correct here. `build-commons.sh` is sourced by other scripts and we really just need openssl. Also, the default directory will be searched anyway, see the pkg-config documentation about the `PKG_CONFIG_PATH` environment variable:
> PKG_CONFIG_PATH
>  A colon-separated  (on  Windows,  semicolon-separated)  list  of
>  directories to search for .pc files.  The default directory will
>  always be searched after searching  the  path;  the  default  is
>  libdir/pkgconfig:datadir/pkgconfig  where  libdir  is the libdir
>  for pkg-config and datadir is the datadir for pkg-config when it
>  was installed.

0xced committed:
Improve cmake error if libssl-dev/openssl is not found

Co-Authored-By: Jeremy Barton <jbarton@microsoft.com>

0xced committed:
Be specific about OpenSSL version 1.1

"Or newer" may not be true

0xced committed:
Remove unnecessary "sh" when running install-native-dependencies.sh

0xced committed:
Revert "Remove unnecessary "sh" when running install-native-dependencies.sh"

This reverts commit 277c5b5ba66c44a0862a9a536e6aec32df747de9.

It turns out the sh was necessary, else we get this error when running on continuous integration:
> /Users/runner/runners/2.166.4/work/_temp/56691a53-a761-472c-b9ab-a9b4eb004f5d.sh: line 1: /Users/runner/runners/2.166.4/work/1/s/eng/install-native-dependencies.sh: Permission denied

0xced committed:
Remove the manual CMake installation instruction

Fixes #34150

0xced commented:
> Since you touching this, can you accommodate #34150 ? Perhaps have brew as the primary option.

Addressed in b2530b11ce057e03e92010a98fb4579823b2a31b. I removed the manual installation instructions altogether, the ICU installation a few lines below only has instruction with Homebrew anyway.

0xced committed:
Merge branch 'master' into simplify-macos-build

safern commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -25,18 +25,12 @@ jobs:
-
     - template: /eng/pipelines/common/clone-checkout-bundle-step.yml
 
+    - ${{ if in(parameters.osGroup, 'OSX', 'iOS', 'tvOS') }}:
```
Merge with the if statement just below?

0xced replied:
Not sure it's a good idea, the conditions are the same but they serve different purposes. Also, *Install Build Dependencies* and *Disk Usage before Build* are already long step names.

safern replied:
What I mean is just use the same condition. Two different steps. Yaml is indentation based so you can have:

```yml
- ${{ if something }}:
  - script: command
     displayName: first command
  - script: command2
     displayName: second command

- script: command3
   displayName: third command
```

So in this example command and command2 are conditioned to something but command3 is unconditional. So what I mean is to move this step after the disk usage step and share the same if, not merge the two commands in one step.

0xced replied:
Got it, done in 43a3d87b96323df90d67326f252eb632c56611ab.

safern commented on eng/pipelines/installer/jobs/base-job.yml:
```diff
@@ -431,6 +431,10 @@ jobs:
         displayName: 'Libraries artifacts (AllConfigurations)'
         cleanUnpackFolder: false
 
+  - ${{ if in(parameters.osGroup, 'OSX', 'iOS', 'tvOS') }}:
```
Merge with the if statement just below?

0xced replied:
Done in 43a3d87b96323df90d67326f252eb632c56611ab.

safern commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -70,13 +70,8 @@ jobs:
-              ln -s /usr/local/opt/openssl/lib/pkgconfig/libcrypto.pc /usr/local/lib/pkgconfig/
-              ln -s /usr/local/opt/openssl/lib/pkgconfig/libssl.pc /usr/local/lib/pkgconfig/
-              ln -s /usr/local/opt/openssl/lib/pkgconfig/openssl.pc /usr/local/lib/pkgconfig/
+        - ${{ if in(parameters.osGroup, 'OSX', 'iOS', 'tvOS') }}:
```
Merge with the if statement for checking disk usage?

0xced replied:
Done in Done in 43a3d87b96323df90d67326f252eb632c56611ab.

0xced committed:
Simplify yaml job files

safern commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -70,19 +70,12 @@ jobs:
-        - ${{ if in(parameters.osGroup, 'OSX', 'iOS','tvOS') }}: 
+        - ${{ if in(parameters.osGroup, 'OSX', 'iOS', 'tvOS') }}:
+          - script: $(Build.SourcesDirectory)/eng/install-native-dependencies.sh ${{ parameters.osGroup }}
+            displayName: Install Build Dependencies
```
Nit: empty line in between tasks, it is easier to read the yml.

0xced replied:
Fixed in c6f19d622e7a713fdb1a2b84fc2ae45137704cb5.

safern commented on eng/pipelines/installer/jobs/base-job.yml:
```diff
@@ -431,7 +431,9 @@ jobs:
-  - ${{ if in(parameters.osGroup, 'OSX', 'iOS','tvOS') }}: 
+  - ${{ if in(parameters.osGroup, 'OSX', 'iOS', 'tvOS') }}:
+    - script: sh $(Build.SourcesDirectory)/eng/install-native-dependencies.sh ${{ parameters.osGroup }}
+      displayName: Install Build Dependencies
```
Nit, empty line for readability.

0xced replied:
Fixed in c6f19d622e7a713fdb1a2b84fc2ae45137704cb5.

safern commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -25,19 +25,11 @@ jobs:
-    - ${{ if in(parameters.osGroup, 'OSX', 'iOS','tvOS') }}: 
+    - ${{ if in(parameters.osGroup, 'OSX', 'iOS', 'tvOS') }}:
+      - script: sh $(Build.SourcesDirectory)/eng/install-native-dependencies.sh ${{ parameters.osGroup }}
+        displayName: Install Build Dependencies
```
Nit, empty line for readability.

0xced replied:
Fixed in c6f19d622e7a713fdb1a2b84fc2ae45137704cb5.

wfurt reviewed (approved):
LGTM. thanks.

safern commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -25,19 +25,12 @@ jobs:
 
-    - ${{ if in(parameters.osGroup, 'OSX', 'iOS','tvOS') }}: 
+    - ${{ if in(parameters.osGroup, 'OSX', 'iOS', 'tvOS') }}:
+      - script: sh $(Build.SourcesDirectory)/eng/install-native-dependencies.sh ${{ parameters.osGroup }}
```
I think you accidentally brought back, `sh` into these steps.

0xced replied:
It was not accidental, see 7037603edce1a0d008685d209a9422beeacdf4ce which reverted the change. All the build were failing in Azure pipelines due to the removal of calling through `sh`.

safern replied:
We have never needed the usage of sh, could you point me to the build error?

0xced replied:
I can't find the corresponding Azure build but here's the error I observed:
> /Users/runner/runners/2.166.4/work/_temp/56691a53-a761-472c-b9ab-a9b4eb004f5d.sh: line 1: /Users/runner/runners/2.166.4/work/1/s/eng/install-native-dependencies.sh: Permission denied

All the jobs running the  `install-native-dependencies.sh` script without running through `sh` would fail with this error. This is why I reverted it in 7037603edce1a0d008685d209a9422beeacdf4ce.

Note that I wrote the scripts to run through `sh` because all other `install-native-dependencies.sh` were run through `sh`, you can observe them on the master branch.

wfurt replied:
```
wfurt-runtime furt$ ls -al eng/install-native-dependencies.sh
-rw-r--r--  1 furt  staff  1293 Apr 28 18:03 eng/install-native-dependencies.sh
```

it seems like that scrips does not have executable bit as it should @safern 
Perhaps something we should fix independently. (or lump here)

safern replied:
Would you like to fix this in a follow up PR?

0xced replied:
@safern I'll leave it up to you.

safern replied:
Ok, it would be nice to fix it (just to be consistent across all places where we invoke bash scripts in the yml), so if you have time that would be great.

0xced commented:
Anything holding the merge?

EgorBo commented:
Oops, it broke android build (when we build it on macOS)
